cohttp-lwt/src/client.ml

```diff
@@ -13,33 +13,23 @@ module Make
   type ctx = Net.ctx [@@deriving sexp_of]
   let default_ctx = Net.default_ctx
 
-  let read_response ~closefn ic _oc meth =
-    Response.read ic >>= begin function
-    | `Invalid reason ->
-      Lwt.fail (Failure ("Failed to read response: " ^ reason))
-    | `Eof -> Lwt.fail (Failure "Client connection was closed")
-    | `Ok res -> begin
-        let has_body = match meth with
-          | `HEAD -> `No
-          | _ -> Response.has_body res
-        in
-        match has_body with
-        | `Yes | `Unknown ->
-          let reader = Response.make_body_reader res ic in
-          let stream = Body.create_stream Response.read_body_chunk reader in
-          let closefn = closefn in
-          Lwt.on_success (Lwt_stream.closed stream) closefn;
-          let gcfn _st = closefn () in
-          Gc.finalise gcfn stream;
-          let body = Body.of_stream stream in
-          Lwt.return (res, body)
-        | `No -> closefn (); Lwt.return (res, `Empty)
-      end
-    end
-    |> fun t ->
-    Lwt.on_cancel t closefn;
-    Lwt.on_failure t (fun _exn -> closefn ());
-    t
+  let read_body ~closefn ic res =
+    match Response.has_body res with
+    | `Yes | `Unknown ->
+      let reader = Response.make_body_reader res ic in
+      let stream = Body.create_stream Response.read_body_chunk reader in
+      let body = Body.of_stream stream in
+      let closed = ref false in
+      Lwt.on_success (Lwt_stream.closed stream)
```

Comment:
That's what I initially suspected because I thougt `Lwt.on_success` will hold a reference on `stream`. But this is not true. It is the other way round. `stream` holds a callback to the `Lwt.on_success` machinery. So the GC will be able to collect `stream`.

Reply:
Oh, I see, thanks. This explains why while testing I was nit noticing any regression.
Can you add a comment about it on this line?

Reply:
It is no longer calling closefn in case of failure or cancellation now though, correct?